File: Java_Advanced/gsapi/src/main/java/br/com/fiap/gs/gsapi/model/Endereco.java
package br.com.fiap.gs.gsapi.model;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "tb_endereco3")
public class Endereco {

    @Id
    @SequenceGenerator(name = "endereco_seq", sequenceName = "tb_endereco3_id_endereco_seq", allocationSize = 1)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "endereco_seq")
    @Column(name = "id_endereco")
    private Long idEndereco;

    @Column(name = "cep", nullable = false, length = 9)
    private String cep;

    @Column(name = "numero", nullable = false)
    private Integer numero; // DDL: NUMBER(5)

    @Column(name = "logradouro", nullable = false, length = 255)
    private String logradouro;

    @Column(name = "bairro", nullable = false, length = 255)
    private String bairro;

    @Column(name = "localidade", nullable = false, length = 100)
    private String localidade;

    @Column(name = "uf", nullable = false, length = 2)
    private String uf;

    @Column(name = "complemento", nullable = false, length = 255)
    private String complemento;

    // CORREÇÃO APLICADA AQUI:
    @Column(name = "latitude", nullable = false, columnDefinition = "NUMBER(10,7)")
    private Double latitude;

    // CORREÇÃO APLICADA AQUI:
    @Column(name = "longitude", nullable = false, columnDefinition = "NUMBER(10,7)")
    private Double longitude;

    @ManyToMany(mappedBy = "enderecos", fetch = FetchType.LAZY)
    private Set<Cliente> clientes = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY,
            cascade = {
                    CascadeType.PERSIST,
                    CascadeType.MERGE
            })
    @JoinTable(name = "tb_enderecoeventos3",
            joinColumns = { @JoinColumn(name = "tb_endereco3_id_endereco") },
            inverseJoinColumns = { @JoinColumn(name = "tb_eonet3_id_eonet") })
    private Set<Eonet> eventosEonet = new HashSet<>();

    // Construtor padrão
    public Endereco() {
    }

    // Construtor para criar um novo endereço
    public Endereco(String cep, Integer numero, String logradouro, String bairro, String localidade, String uf, String complemento, Double latitude, Double longitude) {
        this.cep = cep;
        this.numero = numero;
        this.logradouro = logradouro;
        this.bairro = bairro;
        this.localidade = localidade;
        this.uf = uf;
        this.complemento = complemento;
        this.latitude = latitude;
        this.longitude = longitude;
    }

    // Getters e Setters
    public Long getIdEndereco() {
        return idEndereco;
    }

    public void setIdEndereco(Long idEndereco) {
        this.idEndereco = idEndereco;
    }

    public String getCep() {
        return cep;
    }

    public void setCep(String cep) {
        this.cep = cep;
    }

    public Integer getNumero() {
        return numero;
    }

    public void setNumero(Integer numero) {
        this.numero = numero;
    }

    public String getLogradouro() {
        return logradouro;
    }

    public void setLogradouro(String logradouro) {
        this.logradouro = logradouro;
    }

    public String getBairro() {
        return bairro;
    }

    public void setBairro(String bairro) {
        this.bairro = bairro;
    }

    public String getLocalidade() {
        return localidade;
    }

    public void setLocalidade(String localidade) {
        this.localidade = localidade;
    }

    public String getUf() {
        return uf;
    }

    public void setUf(String uf) {
        this.uf = uf;
    }

    public String getComplemento() {
        return complemento;
    }

    public void setComplemento(String complemento) {
        this.complemento = complemento;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public Set<Cliente> getClientes() {
        return clientes;
    }

    public void setClientes(Set<Cliente> clientes) {
        this.clientes = clientes;
    }

    public Set<Eonet> getEventosEonet() {
        return eventosEonet;
    }

    public void setEventosEonet(Set<Eonet> eventosEonet) {
        this.eventosEonet = eventosEonet;
    }

    public void addEventoEonet(Eonet evento) {
        this.eventosEonet.add(evento);
    }

    public void removeEventoEonet(Eonet evento) {
        this.eventosEonet.remove(evento);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Endereco endereco = (Endereco) o;
        return Objects.equals(idEndereco, endereco.idEndereco);
    }

    @Override
    public int hashCode() {
        return Objects.hash(idEndereco);
    }

    @Override
    public String toString() {
        return "Endereco{" +
                "idEndereco=" + idEndereco +
                ", cep='" + cep + '\'' +
                ", numero=" + numero +
                ", logradouro='" + logradouro + '\'' +
                ", bairro='" + bairro + '\'' +
                ", localidade='" + localidade + '\'' +
                ", uf='" + uf + '\'' +
                ", complemento='" + complemento + '\'' +
                ", latitude=" + latitude +
                ", longitude=" + longitude +
                '}';
    }
}

File: Advanced_Business_Development_with.NET/gsApi/gsApi/model/Endereco.cs
// File: gsApi/model/Endereco.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace gsApi.model
{
    [Table("TB_ENDERECO3")]
    public class Endereco
    {
        [Key]
        [Column("ID_ENDERECO")]
        public long IdEndereco { get; set; }

        [Required(ErrorMessage = "O CEP não pode estar em branco.")]
        [StringLength(9)]
        [Column("CEP")]
        public required string Cep { get; set; }

        [Required(ErrorMessage = "O número do endereço é obrigatório.")]
        [Column("NUMERO")]
        public int Numero { get; set; }

        [Required(ErrorMessage = "O logradouro não pode estar em branco.")]
        [StringLength(255)]
        [Column("LOGRADOURO")]
        public required string Logradouro { get; set; }

        [Required(ErrorMessage = "O bairro não pode estar em branco.")]
        [StringLength(255)]
        [Column("BAIRRO")]
        public required string Bairro { get; set; }

        [Required(ErrorMessage = "A localidade (cidade) não pode estar em branco.")]
        [StringLength(100)]
        [Column("LOCALIDADE")]
        public required string Localidade { get; set; }

        [Required(ErrorMessage = "A UF não pode estar em branco.")]
        [StringLength(2)]
        [Column("UF")]
        public required string Uf { get; set; }

        [Required(ErrorMessage = "O complemento é obrigatório.")] // Mantido como required conforme modelo original
        [StringLength(255)]
        [Column("COMPLEMENTO")]
        public required string Complemento { get; set; }

        [Required(ErrorMessage = "Latitude é obrigatória.")]
        [Column("LATITUDE", TypeName = "NUMBER(10,7)")]
        public double Latitude { get; set; }

        [Required(ErrorMessage = "Longitude é obrigatória.")]
        [Column("LONGITUDE", TypeName = "NUMBER(10,7)")]
        public double Longitude { get; set; }

        [JsonIgnore]
        public virtual ICollection<Cliente> Clientes { get; set; } = new List<Cliente>();
        public virtual ICollection<EonetEvent> EventosEonet { get; set; } = new List<EonetEvent>();
    }
}